nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,372 @@
+        mask_frame_, global_frame_, tf2::TimePointZero,
+        transform_tolerance_);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR(
+        node->get_logger(),
+        "SpeedFilter: failed to get costmap frame (%s) "
+        "transformation to mask frame (%s) with error: %s",
+        global_frame_.c_str(), mask_frame_.c_str(), ex.what());
+      return false;
+    }
+    tf2::fromMsg(transform.transform, tf2_transform);
+
+    // Transform (pose.x, pose.y) point from global_frame_
+    // to mask_pose point in mask_frame_
+    tf2::Vector3 point(pose.x, pose.y, 0);
+    point = tf2_transform * point;
+    mask_pose.x = point.x();
+    mask_pose.y = point.y();
+  } else {
+    // Filter mask and current layer are in the same frame:
+    // Just use pose coordinates
+    mask_pose = pose;
+  }
+
+  return true;
+}
+
+bool SpeedFilter::worldToMask(double wx, double wy, unsigned int & mx, unsigned int & my) const
+{
+  double origin_x = filter_mask_->info.origin.position.x;
+  double origin_y = filter_mask_->info.origin.position.y;
+  double resolution = filter_mask_->info.resolution;
+  unsigned int size_x = filter_mask_->info.width;
+  unsigned int size_y = filter_mask_->info.height;
+
+  if (wx < origin_x || wy < origin_y) {
+    return false;
+  }
+
+  mx = std::round((wx - origin_x) / resolution);
+  my = std::round((wy - origin_y) / resolution);
+  if (mx >= size_x || my >= size_y) {
+    return false;
+  }
+
+  return true;
+}
+
+void SpeedFilter::process(
+  nav2_costmap_2d::Costmap2D & /*master_grid*/,
+  int /*min_i*/, int /*min_j*/, int /*max_i*/, int /*max_j*/,
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!filter_mask_) {
+    // Show warning message every 2 seconds to not litter an output
+    RCLCPP_WARN_THROTTLE(
+      node->get_logger(), *(node->get_clock()), 2000,
+      "Filter mask was not received");
+    return;
+  }
+
+  geometry_msgs::msg::Pose2D mask_pose;  // robot coordinates in mask frame
+
+  // Transforming robot pose from current layer frame to mask frame
+  if (!transformPose(pose, mask_pose)) {
+    // Robot pose is unknown. Returning.
+    return;
+  }
+
+  // Converting mask_pose robot position to filter_mask_ indexes (mask_robot_i, mask_robot_j)
+  unsigned int mask_robot_i, mask_robot_j;
+  if (!worldToMask(mask_pose.x, mask_pose.y, mask_robot_i, mask_robot_j)) {
+    // Robot coordinates are out of mask bounds. Returning.
+    return;
+  }
+
+  // Getting filter_mask data from cell where the robot placed and
+  // calculating speed limit value
+  unsigned int size_x = filter_mask_->info.width;
+  int8_t data = filter_mask_->data[mask_robot_j * size_x + mask_robot_i];
+  if (data != nav2_util::OCC_GRID_FREE && data != nav2_util::OCC_GRID_UNKNOWN) {
```

Comment:
This is `OccupancyGrid` constants, not a `Costmap2D` ones:
```
$ cat nav2_util/include/nav2_util/occ_grid_values.hpp
...
namespace nav2_util
{
static constexpr int8_t OCC_GRID_UNKNOWN = -1;
static constexpr int8_t OCC_GRID_FREE = 0;
static constexpr int8_t OCC_GRID_OCCUPIED = 100;
}
```
They are both applied to data of `OccupancyGrid` type, so I think this should be the right usage.

Reply:
`filter_mask_` is our data representing the hand-annotated speed information. While `filter_mask_` is a costmap type, it is not a costmap / occupancy grid (its a "speed grid" I suppose :smile: ). So applying constant values meant to represent occupancy values doesn't really make much since to a "speed grid" that doesn't contain occupancy information. 0 on the speed filter mask is a totally valid value - its 0% max speed with default base/multiplier values, or anything really if `base != 0`. 100 is also incredibly value as the default for 100% speed with default base/multiplier values.

Reply:
Okay, looks like this should be the same constants, but named in other abstract space (because of `OccupancyGrid` mask for `SpeedFilter` is not the map in its usual understanding). I will duplicate these constants and change their naming to more representative (say `SPEED_MASK_FREE`, `SPEED_MASK_UNKNOWN`) made for SpeedFilter's OccupancyGrid in next update.

Reply:
Introduced new constants: `SPEED_MASK_UNKNOWN` and `SPEED_MASK_NO_LIMIT` in https://github.com/ros-planning/navigation2/commit/3465c90415afd11418f98fc01bbc9386afa853ea#diff-6f9b5b73d858cc4570fd2d807b4ab6ad2c43be02da2ec6f34db8fac5708b9a50R56-R57

Reply:
What situation is `SPEED_MASK_UNKNOWN` used? I'm not sure that's sensible, what should we do in this situation? Basically just use `NO_LIMIT`? I think that can be removed.

Reply:
I've described my doubts about complete removal of `SPEED_MASK_UNKNOWN` constant in https://github.com/ros-planning/navigation2/pull/2074#discussion_r535928710.